aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -368,3 +458,101 @@ impl<'r, 'l> DerefMut for SessionExt<'r, 'l> {
+        legacy_creation_as_modification: bool,
+    ) -> Result<WriteOp, VMStatus> {
+        use MoveStorageOp::*;
+        use WriteOp::*;
+
+        let existing_value_opt = self.remote.get_state_value(state_key).map_err(|_| {
+            VMStatus::Error(
+                StatusCode::STORAGE_ERROR,
+                err_msg("Storage read failed when converting change set."),
+            )
+        })?;
+
+        let write_op = match (existing_value_opt, move_storage_op) {
+            (None, Modify(_) | Delete) => {
+                return Err(VMStatus::Error(
+                    StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                    err_msg("When converting write op: updating non-existent value."),
+                ))
+            },
+            (Some(_), New(_)) => {
+                return Err(VMStatus::Error(
+                    StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                    err_msg("When converting write op: Recreating existing value."),
+                ))
+            },
+            (None, New(data)) => match &self.new_slot_metadata {
+                None => {
+                    if legacy_creation_as_modification {
+                        Modification(data)
+                    } else {
+                        Creation(data)
+                    }
+                },
+                Some(metadata) => CreationWithMetadata {
+                    data,
+                    metadata: metadata.clone(),
+                },
+            },
+            (Some(existing_value), Modify(data)) => {
+                // Inherit metadata even if the feature flags is turned off, for compatibility.
+                match existing_value.into_metadata() {
+                    None => Modification(data),
+                    Some(metadata) => ModificationWithMetadata { data, metadata },
+                }
+            },
+            (Some(existing_value), Delete) => {
+                // Inherit metadata even if the feature flags is turned off, for compatibility.
+                match existing_value.into_metadata() {
+                    None => Deletion,
+                    Some(metadata) => DeletionWithMetadata { metadata },
+                }
+            },
+        };
+        Ok(write_op)
+    }
+
+    fn convert_aggregator_mod(
+        &self,
+        state_key: &StateKey,
+        value: u128,
+    ) -> Result<WriteOp, VMStatus> {
+        let existing_value_opt = self
+            .remote
+            .get_state_value(state_key)
+            .map_err(|_| VMStatus::Error(StatusCode::STORAGE_ERROR, None))?;
+        let data = serialize(&value);
+
+        let op = match existing_value_opt {
+            None => {
+                match &self.new_slot_metadata {
+                    // n.b. Aggregator writes historically did not distinguish Create vs Modify.
+                    None => WriteOp::Modification(data),
+                    Some(metadata) => WriteOp::CreationWithMetadata {
```

Comment:
What will happen if multiple "creates" from different txns that have different metadata?

Reply:
This is after the aggregators are materialized, there shouldn't be multiple creates between txns.

Reply:
I don't think aggregator creation will be here iirc, aggregator creation should go to normal writeset. i.e. None on existing_value should be impossible?

also the word "materialized" is confusing, I think you mean after aggregators are created.

Reply:
This function is to replace:
https://github.com/aptos-labs/aptos-core/pull/7600#discussion_r1190338606

And this special treatment fixed tests `None => WriteOp::Modification(data),`

let me remove this and see again.